Test: wallet issue with orphaned rewards

domob1812 opened the pull request:
This adds a new test case demonstrating the wallet issue when block rewards are orphaned (#14148).

domob1812 committed:
Testcase for wallet issue with orphaned rewards.

This adds a new test case demonstrating the wallet issue when block
rewards are orphaned (https://github.com/bitcoin/bitcoin/issues/14148).

domob1812 commented:
This is meant as an example for #14148, not to be merged for now.

laanwj commented:
If this demonstrates an unfixed issue, shouldn't the test fail on master?

domob1812 commented:
> If this demonstrates an unfixed issue, shouldn't the test fail on master?

That's a good point.  The test was originally written to assert the current behaviour and thus passed (but looking at the code one could see the behaviour described in #14148).

But I agree that is probably confusing.  I've added another commit now that makes the test fail as expected (the original version is still preserved in the first commit, as it might be interesting as well for understanding what the issue is about).

MarcoFalke commented on test/functional/wallet_orphanedreward.py:
```diff
@@ -0,0 +1,46 @@
+        self.nodes[0].invalidateblock(blk)
+        self.nodes[0].generate(200)
+        self.sync_blocks()
+        self.nodes[1].sendtoaddress(self.nodes[0].getnewaddress(), 9)
```
Could comment this out, and add a comment that it fails?

domob1812 replied:
Instead of this, I've now added in the `abandontransaction` call that makes the test succeed (but according to #14148 should not be needed) as well as a comment explaining this.  I think that makes the situation even more clear.

MarcoFalke commented on test/functional/wallet_orphanedreward.py:
```diff
@@ -0,0 +1,46 @@
+        self.nodes[0].invalidateblock(blk)
+        self.nodes[0].generate(200)
+        self.sync_blocks()
+        self.nodes[1].sendtoaddress(self.nodes[0].getnewaddress(), 9)
```
Also, would be good to check the output of `getbalances` here.

domob1812 replied:
Added.

maflcko reviewed (commented):
Concept ACK

domob1812 commented on test/functional/wallet_orphanedreward.py:
```diff
@@ -0,0 +1,46 @@
+"""Test orphaned block rewards in the wallet."""
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal, assert_raises_rpc_error
```
Indeed, that's a left over from the previous commit.  Removed.

domob1812 commented on test/functional/wallet_orphanedreward.py:
```diff
@@ -0,0 +1,46 @@
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 2
+        self.supports_cli = False
```
Removed.

domob1812 commented on test/functional/wallet_orphanedreward.py:
```diff
@@ -0,0 +1,46 @@
+
+class OrphanedBlockRewardTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
```
I think that a clean chain makes the test easier to read, as it doesn't require any prior knowledge about how the cached chain looks like.  If we use the cached chain, it will be harder to make sure that `node[1]` only has the coins we want to, as it will have immature coins that will appear over time.

Also, it seems the initial setup (including until after the `sendtoaddress`) is quick anyway.  The slow parts are the `sync_blocks` calls, which we wouldn't get rid of with a cached chain.

Or did I misunderstand your suggestion?

domob1812 commented:
Thanks for the feeback, @MarcoFalke !  I've addressed it with a new commit (so as to not invalidate comments made on previous commits), but am of course happy to squash all history into a single commit any time.

maflcko reviewed (approved):
🄰🄲🄺 I think I am going to merge this.

Maybe squash the commits and add a check for `getbalances` also *before* the abandontransaction call. Due to the bug it should be all zeros, right?

domob1812 commented:
@MarcoFalke Thanks!  I've rebased onto latest `master`, squashed all commits, and added in a `getbalances` check before the `abandontransaction` call as well (which indeed shows all zero).

MarcoFalke commented:
ACK 7aeaa92ae0d45cddd2cdbdc735cdbcfa5fb2871a

decryp2kanon commented:
Concept ACK

LarryRuane commented on test/functional/wallet_orphanedreward.py:
```diff
@@ -0,0 +1,61 @@
+        # Orphan the block reward and make sure that the original coins
+        # from the wallet can still be spent.
+        self.nodes[0].invalidateblock(blk)
+        self.nodes[0].generate(200)
```
I wasn't sure why this `generate(200)` was needed, so I commented it out, and the test hung in the next line (`sync_blocks()`). I restored the line and played around with different generate counts, and found that it fails with 151 and works at 152. If I'm counting correctly, we invalidated 151 blocks, so it seems that it's necessary to extend the chain so it's _longer_ than it was before. (Necessary for this test framework, not in real life.)

You could reduce this to 152; I like tests to "just barely" pass because it seems like that imparts better understanding to readers (I wondered is, "why is this 200?"), and also because if the code is acting differently from our understanding, for example, we think generating 152 blocks should work but it actually takes 170, and the test does 200, then that mistaken understanding (or even a bug in the code) would be papered over. 

Also, the test runs a little faster if we reduce this.

Of course, a downside to tests "just barely" passing is that they're more fragile. It's painful when someone makes a perfectly good change and one of the tests break! (I don't think there's much of a chance of that happening here.)

domob1812 replied:
Sounds good to me, updated to 152.

LarryRuane commented on test/functional/wallet_orphanedreward.py:
```diff
@@ -0,0 +1,61 @@
+          "untrusted_pending": 0,
+          "immature": 0,
+        })
+        self.nodes[1].sendtoaddress(address=self.nodes[0].getnewaddress(), amount=9, subtractfeefromamount=True)
```
Can this be removed?

domob1812 replied:
Yes it can, but I think having it in there is an extra check that the wallet does properly handle the coins.  So unless there is a strong preference to remove it, I'd rather keep it in to make the test more robust and descriptive.

LarryRuane replied:
> ... having it in there is an extra check ...

I verified that if this `sendtoaddress` fails (which I forced by increasing the amount sent to 99999), the entire test fails, which is good. (I wasn't aware of that, which is why I thought this was an unnecessary line of code -- good to leave it in as you have. Need to put an RPC within a `try` to have its failure not fail the overall test.)

LarryRuane reviewed (commented):
ACK 7aeaa92ae0d45cddd2cdbdc735cdbcfa5fb2871a
Nice PR! I made sure the test passes as written, and that removing the `abandontransaction` makes it fail as expected. Suggestions are minor nits, nonblocking.

leonardojobim reviewed (approved):
tACK https://github.com/bitcoin/bitcoin/pull/18795/commits/7aeaa92ae0d45cddd2cdbdc735cdbcfa5fb2871a. Tested on Ubuntu 20.04.

The test successfully demonstrates that when an orphaned block reward is spent together with other outputs, the wallet stops showing any funds and that "abandontransaction(txid)" restores the valid funds.


Agree with https://github.com/bitcoin/bitcoin/pull/18795#discussion_r544686136 and https://github.com/bitcoin/bitcoin/pull/18795#discussion_r544686860. Using 152 blocks apparently works fine and there's no need of the last line.

The code below would also work as the final validation of this test:
```
        self.stop_node(1)
        self.start_node(1, ["-zapwallettxes=2"])

        assert_equal(self.nodes[1].getbalances()["mine"], {
          "trusted": 10,
          "untrusted_pending": 0,
          "immature": 0,
        })
```

domob1812 commented on test/functional/wallet_orphanedreward.py:
```diff
@@ -0,0 +1,61 @@
+        # the existing balance and the block reward.
+        self.nodes[0].generate(150)
+        assert_equal(self.nodes[1].getbalance(), 10 + 25)
+        txid = self.nodes[1].sendtoaddress(address=self.nodes[0].getnewaddress(), amount=30, subtractfeefromamount=True)
```
Indeed.  Not sure why I had the other form; maybe a previous version of the test needed the `subtractfeefromamount`, but that is no longer the case.

domob1812 commented:
Thanks @LarryRuane and @leonardojobim , and sorry for the long response time on my end.  I have now rebased the PR onto latest `master` and addressed your comments.

As for the alternate final validation with `-zapwallettxes`, I agree this is possible, but I think `abandontransaction` is more explicit as it specifies the transaction in question.  I'm happy to change it, though.

LarryRuane commented:
ACK ef9f3f4b25a6158aa3ff32bf74c104c325c910b1

leonardojobim reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/18795/commits/ef9f3f4b25a6158aa3ff32bf74c104c325c910b1

adamjonas commented:
Ping for rebase @domob1812.

domob1812 commented:
Thanks for the reminder, rebased.

LarryRuane commented on test/functional/wallet_orphanedreward.py:
```diff
@@ -0,0 +1,61 @@
+        self.nodes[0].sendtoaddress(self.nodes[1].getnewaddress(), 10)
+        self.nodes[0].generate(1)
+
+        # Get a block reward with node 1 and remember the block so we can orphan
```
```suggestion
        # Get a block reward (25) with node 1 and remember the block so we can orphan
```
(Non-blocking, only if you touch-up, would have made the test easier for me to understand.)

LarryRuane reviewed (approved):
ACK e4356f6a6c18e5027a064a4d3a5deda27985f584
I verified that without the `abandontransaction`, the assertion fails, and (if I comment out the assertion) the `sendtoaddress` fails.

adamjonas commented:
CI fail is unrelated, see https://github.com/bitcoin/bitcoin/issues/20975

leonardojobim reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/18795/commits/e4356f6a6c18e5027a064a4d3a5deda27985f584 .
Tested on Ubuntu 20.04.
